Core/src/GapicRequestWrapper.php

```diff
@@ -0,0 +1,189 @@
+        } catch (\Exception $ex) {
+            if ($ex instanceof ApiException) {
+                throw $this->convertToGoogleException($ex);
+            }
+
+            throw $ex;
+        }
+    }
+
+    /**
+     * Serializes a gRPC response.
+     *
+     * @param mixed $response
+     * @return \Generator|OperationResponse|array|null
+     */
+    private function handleResponse($response)
+    {
+        if ($response instanceof PagedListResponse) {
+            $response = $response->getPage()->getResponseObject();
+        }
+
+        if ($response instanceof Message) {
+            return $this->serializer->encodeMessage($response);
+        }
+
+        if ($response instanceof OperationResponse) {
+            return $response;
+        }
+
+        if ($response instanceof ServerStream) {
+            return $this->handleStream($response);
+        }
+
+        return null;
+    }
+
+    /**
+     * Handles a streaming response.
+     *
+     * @param ServerStream $response
+     * @return \Generator|array|null
+     * @throws Exception\ServiceException
+     */
+    private function handleStream($response)
+    {
+        try {
+            foreach ($response->readAll() as $count => $result) {
+                $res = $this->serializer->encodeMessage($result);
+                yield $res;
+            }
+        } catch (\Exception $ex) {
+            throw $this->convertToGoogleException($ex);
+        }
+    }
+
+    /**
+     * Convert a ApiCore exception to a Google Exception.
+     *
+     * @param \Exception $ex
+     * @return Exception\ServiceException
+     */
+    private function convertToGoogleException($ex)
+    {
+        switch ($ex->getCode()) {
+            case Code::INVALID_ARGUMENT:
+                $exception = Exception\BadRequestException::class;
+                break;
+
+            case Code::NOT_FOUND:
+            case Code::UNIMPLEMENTED:
+                $exception = Exception\NotFoundException::class;
+                break;
+
+            case Code::ALREADY_EXISTS:
+                $exception = Exception\ConflictException::class;
+                break;
+
+            case Code::FAILED_PRECONDITION:
+                $exception = Exception\FailedPreconditionException::class;
+                break;
+
+            case Code::UNKNOWN:
+                $exception = Exception\ServerException::class;
+                break;
+
+            case Code::INTERNAL:
+                $exception = Exception\ServerException::class;
+                break;
+
+            case Code::ABORTED:
+                $exception = Exception\AbortedException::class;
+                break;
+
+            case Code::DEADLINE_EXCEEDED:
+                $exception = Exception\DeadlineExceededException::class;
+                break;
+
+            default:
+                $exception = Exception\ServiceException::class;
+                break;
+        }
+
+        $metadata = [];
+        if ($ex->getMetadata()) {
```

Comment:
I don't think we can have the typehint in this method as we would need a union type, because I see a place which calls `convertToGoogleException` with a generic `Exception` and not `ApiException`.

I have used `method_exists` to cater this.